src/Host/Client/Impl/Session/RSessionProvider.cs

```diff
@@ -206,37 +212,109 @@ public class RSessionProvider : IRSessionProvider {
             return true;
         }
 
-        private async Task<bool> TrySwitchSessionsAsync(List<RSession> sessions, IBrokerClient oldBroker, CancellationToken cancellationToken) {
+        private async Task SwitchBrokerAsync(CancellationToken cancellationToken, IBrokerClient oldBroker) {
+            var switchingFromNull = oldBroker is NullBrokerClient;
+            if (!switchingFromNull) {
+                _callback.WriteConsole(Resources.RSessionProvider_StartSwitchingWorkspaceFormat.FormatInvariant(_brokerProxy.Name, GetUriString(_brokerProxy)));
+            }
+
+            var sessions = _sessions.Values.ToList();
+            if (sessions.Any()) {
+                await SwitchSessionsAsync(sessions, oldBroker, cancellationToken);
+            } else {
+                // Ping isn't enough here - need a "full" test with RHost
+                await TestBrokerConnectionWithRHost(_brokerProxy, cancellationToken);
+            }
+
+            if (!switchingFromNull) {
+                _callback.WriteConsole(Resources.RSessionProvider_SwitchingRWorkspaceCompleted);
+            }
+        }
+
+        private async Task SwitchSessionsAsync(IEnumerable<RSession> sessions, IBrokerClient oldBroker, CancellationToken cancellationToken) {
+            var cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+
             // All sessions should participate in switch. If any of it didn't start, cancel the rest.
-            var startTransactionTasks = sessions.Select(s => s.StartSwitchingBrokerAsync(cancellationToken)).ToList();
+            var startTransactionTasks = sessions.Select(s => s.StartSwitchingBrokerAsync(cts.Token)).ToList();
+
             try {
                 await Task.WhenAll(startTransactionTasks);
-            } catch (OperationCanceledException) {
+                var transactions = startTransactionTasks.Select(t => t.Result).ToList();
+
+                _callback.WriteConsole(Resources.RSessionProvider_StartConnectingToWorkspaceFormat.FormatInvariant(transactions.Count));
+                await Task.WhenAll(transactions.Select(t => ConnectToNewBrokerAsync(t, cts)));
+
+                _callback.WriteConsole(Resources.RSessionProvider_RestartingSessionsFormat.FormatInvariant(transactions.Count));
+                OnBrokerDisconnected();
+                await Task.WhenAll(transactions.Select(t => CompleteSwitchingBrokerAsync(t, oldBroker, cts)));
+            } finally {
                 foreach (var task in startTransactionTasks.Where(t => t.Status == TaskStatus.RanToCompletion)) {
```

Comment:
Without `Where`, `task.Result` will throw exception on the tasks that have failed.

---
In reply to: [80110049](https://github.com/Microsoft/RTVS/pull/2384#discussion_r80110049) [](ancestors = 80110049)